core/domain/stats_services.py

```diff
@@ -212,6 +213,165 @@ def handle_stats_creation_for_new_exp_version(
     create_stats_model(exploration_stats)
 
 
+def create_exp_issues_for_new_exploration(exp_id, exp_version):
+    """Creates the ExplorationIssuesModel instance for the exploration.
+
+    Args:
+        exp_id: str. ID of the exploration.
+        exp_version: int. Version of the exploration.
+    """
+    stats_models.ExplorationIssuesModel.create(exp_id, exp_version, [])
+
+
+def update_exp_issues_for_new_exp_version(exploration, change_list):
+    """Retrieves the ExplorationIssuesModel for the old exp_version and makes
+    any required changes to the structure of the model.
+
+    Args:
+        exploration: Exploration. Domain object for the exploration.
+        change_list: list(dict). A list of changes introduced in this commit.
+    """
+    # TODO(pranavsid98): Convert below two lines to use get_by_id() instead.
+    exp_issues_model = stats_models.ExplorationIssuesModel.get_model(
+        exploration.id, exploration.version - 1)
+    if exp_issues_model is None:
+        create_exp_issues_for_new_exploration(
+            exploration.id, exploration.version - 1)
+        return
+    exp_issues = get_exp_issues_from_model(exp_issues_model)
+
+    # We make a mapping from state name to list of exploration issues so that
+    # handling those issues will be easier while going through the change list.
+    exp_issues_by_state_name = collections.defaultdict(list)
+    for exp_issue in exp_issues.unresolved_issues:
+        keyname = stats_models.ISSUE_TYPE_KEYNAME_MAPPING[exp_issue.issue_type]
+        # If keyname is 'state_names', which is a list, we add an instance of
+        # this issue to all associated states in the mapping.
+        if keyname == 'state_names':
+            state_names = exp_issue.issue_customization_args[keyname]['value']
+            for state_name in state_names:
+                exp_issues_by_state_name[state_name].append(exp_issue)
+        else:
+            state_name = exp_issue.issue_customization_args[keyname]['value']
+            exp_issues_by_state_name[state_name].append(exp_issue)
+
+    exp_versions_diff = exp_domain.ExplorationVersionsDiff(change_list)
+
+    # Handling deletions.
+    for state_name in exp_versions_diff.deleted_state_names:
+        for exp_issue in exp_issues_by_state_name[state_name]:
+            idx = exp_issues.unresolved_issues.index(exp_issue)
+            exp_issues.unresolved_issues[idx].is_valid = False
```

Comment:
If an issue becomes invalid, what happens to its resolution status? If it stays in "unresolved issues" then how does it ever get resolved? Is it by manual creator action?

Reply:
@seanlip It'll still stay in unresolved issues. It can only get resolved if there is a revert at any point, at which time, the issue becomes valid again and the creator resolves it then.

Reply:
Ah OK. But unresolved invalid issues won't be shown to the creator in the UI? They'll just disappear? (Would that be ... weird?)

Maybe better to still keep them there but have a mark on them saying "resolved"? But then if these never get removed from the creator's "todo list", then that isn't so great since the creator can't do anything about them. (For that matter, how are we planning to represent resolved issues in the UI?)

So, overall, I guess I'm asking --- what exactly does the creator experience look like here?

Reply:
@seanlip Unresolved invalid issues won't be shown to the creator, yes. Well, its basically because these playthroughs don't make any more sense. If the creator has deleted a state, and a bunch of learners had an EarlyQuit issue at that same state, we can't go around showing those playthroughs right?

But, this brings up another thought. State deletions are the only way we are invalidating playthroughs. Interestingly, deleting a state is one way of resolving something no? Learners have issues in a state, and that state is deleted now. This brings up a possibility of resolving all issues with concerned state as deleted states and do away the validity altogether.

What do you think?

Reply:
Yes, I think that seems like a reasonable interpretation. I don't have objections to this, but I haven't thought about it deeply. I think @brianrodri and you would be in a better position to decide.

@brianrodri, any thoughts?

Reply:
I do see value in keeping the states around. We already have the logic to "revert" an exploration through a sequence of changes. It follows that a playthrough can remain relevant, but only if we perform those same reverting steps on the exploration first.

I also disagree that deleting a state == resolving it entirely. An issue arose because of a learner's misunderstanding of the lesson, not necessarily because of the particular state which was deleted. If anything, the state which a learner will quit will likely shift, not disappear entirely.

Given that, I believe we should keep old playthroughs around, and allow creators to look at them by applying the "revert" steps onto the exploration before taking a look at them. In this case, I think it'd be more useful to keep track of the exploration_version instead of an "is_valid" bit, then only show (for now) the latest exploration versions to the creators.

Reply:
@brianrodri I wouldn't necessarily assume it would need to shift though right? If a learner was able to come up until that state and quit there, without the state, he logically moves forward. But you are right, it doesn't mean that the issue needs to have been resolved. Ultimately, I don't mind keeping the older playthroughs around and show them when a revert happens - it seems to be better engineering.

We already keep track of the exploration version for ExplorationIssuesModel. That's how we are handling reverts effectively now.